BUG: With cache, to_datetime() returns pd.NaT for inputs that produce duplicated values

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
from pandas import NaT, to_datetime, Series, Timestamp
from pandas._testing import assert_series_equal

input_ser = Series([None] + [NaT] * 50 + ["2012 July 26", Timestamp("2012-07-26")], dtype="object")
expected_ser = Series([NaT] * 51 + [Timestamp("2012-07-26"), Timestamp("2012-07-26")], dtype="datetime64[ns]")
result = to_datetime(input_ser)
assert_series_equal(result, expected_ser)
# AssertionError: numpy array are different

# actual result is [NaT] * 51 + [Timestamp("2012-07-26"), NaT]
# ie. below passes
# assert_series_equal(result, Series([NaT] * 51 + [Timestamp("2012-07-26"), NaT], dtype="datetime64[ns]")) 
```

#### Problem description

The current to_datetime will incorrectly parse and omit data in certain situations due to a slightly erroneous deduplication to fix
GH#39882 and GH#35888.

#### Expected Output

to parse the datetime correctly. 
Eg. return Series([NaT] * 51 + [Timestamp("2012-07-26"), Timestamp("2012-07-26")], dtype="datetime64[ns]") in the example above.

#### Output of ``pd.show_versions()``

<details>

pandas           : 1.4.0.dev0+108.gfa6b96e128
numpy            : 1.21.0
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.2
setuptools       : 57.0.0
Cython           : 0.29.23
pytest           : 6.2.4
hypothesis       : 6.14.0
sphinx           : 4.0.2
blosc            : 1.10.4
feather          : None
xlsxwriter       : 1.4.3
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.25.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 2021.05.0
fastparquet      : 0.6.3
gcsfs            : 2021.05.0
matplotlib       : 3.4.2
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : 4.0.1
pyxlsb           : None
s3fs             : 2021.05.0
scipy            : 1.7.0
sqlalchemy       : 1.4.19
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1

</details>

--- a/doc/source/whatsnew/v1.4.0.rst
+++ b/doc/source/whatsnew/v1.4.0.rst
@@ -122,7 +122,7 @@ Categorical
 
 Datetimelike
 ^^^^^^^^^^^^
--
+- Bug in :func:`to_datetime` returning pd.NaT for inputs that produce duplicated values, when ``cache=True`` (:issue:`42259`)
 -
 
 Timedelta

--- a/pandas/core/tools/datetimes.py
+++ b/pandas/core/tools/datetimes.py
@@ -194,9 +194,9 @@ def _maybe_cache(
         if len(unique_dates) < len(arg):
             cache_dates = convert_listlike(unique_dates, format)
             cache_array = Series(cache_dates, index=unique_dates)
-            if not cache_array.is_unique:
+            if not cache_array.index.is_unique:
                 # GH#39882 in case of None and NaT we get duplicates
-                cache_array = cache_array.drop_duplicates()
+                cache_array = cache_array[~cache_array.index.unique()]
     return cache_array
 
 

--- a/pandas/core/tools/datetimes.py
+++ b/pandas/core/tools/datetimes.py
@@ -194,10 +194,9 @@ def _maybe_cache(
         if len(unique_dates) < len(arg):
             cache_dates = convert_listlike(unique_dates, format)
             cache_array = Series(cache_dates, index=unique_dates)
+            # GH#39882 and GH#35888
             if not cache_array.index.is_unique:
-                cache_array = cache_array[
-                    ~cache_array.index.duplicated()
-                ]
+                cache_array = cache_array[~cache_array.index.duplicated()]
     return cache_array
 
 

--- a/pandas/core/tools/datetimes.py
+++ b/pandas/core/tools/datetimes.py
@@ -195,8 +195,9 @@ def _maybe_cache(
             cache_dates = convert_listlike(unique_dates, format)
             cache_array = Series(cache_dates, index=unique_dates)
             if not cache_array.index.is_unique:
-                # GH#39882 in case of None and NaT we get duplicates
-                cache_array = cache_array[~cache_array.index.unique()]
+                cache_array = cache_array[
+                    ~cache_array.index.duplicated()
+                ]
     return cache_array
 
 

--- a/pandas/core/tools/datetimes.py
+++ b/pandas/core/tools/datetimes.py
@@ -194,7 +194,7 @@ def _maybe_cache(
         if len(unique_dates) < len(arg):
             cache_dates = convert_listlike(unique_dates, format)
             cache_array = Series(cache_dates, index=unique_dates)
-            # GH#39882 and GH#35888
+            # GH#39882 and GH#35888 in case of None and NaT we get duplicates
             if not cache_array.index.is_unique:
                 cache_array = cache_array[~cache_array.index.duplicated()]
     return cache_array

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -969,10 +969,11 @@ class TestToDatetime:
                 (NaT, Timestamp("2012-07-26")) + (NaT,) * start_caching_at,
             ),
             (
-                (None, "2012 July 26", Timestamp("2012-07-26"))
-                + (NaT,) * start_caching_at,
-                (NaT, Timestamp("2012-07-26"), Timestamp("2012-07-26"))
-                + (NaT,) * start_caching_at,
+                (None,)
+                + (NaT,) * start_caching_at
+                + ("2012 July 26", Timestamp("2012-07-26")),
+                (NaT,) * (start_caching_at + 1)
+                + (Timestamp("2012-07-26"), Timestamp("2012-07-26")),
             ),
         ),
     )

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -8,7 +8,6 @@ from datetime import (
 )
 from decimal import Decimal
 import locale
-from typing import Any, List, Union
 
 from dateutil.parser import parse
 from dateutil.tz.tz import tzoffset
@@ -977,7 +976,9 @@ class TestToDatetime:
             ),
         ),
     )
-    def test_convert_object_to_datetime_with_cache(self, datetimelikes, expected_values):
+    def test_convert_object_to_datetime_with_cache(
+        self, datetimelikes, expected_values
+    ):
         # GH#39882
         ser = Series(
             datetimelikes,

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -959,7 +959,7 @@ class TestToDatetime:
         assert result == expected
 
     @pytest.mark.parametrize(
-        "datetimelikes,expected",
+        "datetimelikes,expected_values",
         (
             (
                 (None, np.nan) + (NaT,) * start_caching_at,
@@ -977,7 +977,7 @@ class TestToDatetime:
             ),
         ),
     )
-    def test_convert_object_to_datetime_with_cache(self, datetimelikes, expected):
+    def test_convert_object_to_datetime_with_cache(self, datetimelikes, expected_values):
         # GH#39882
         ser = Series(
             datetimelikes,
@@ -985,7 +985,7 @@ class TestToDatetime:
         )
         result_series = to_datetime(ser, errors="coerce")
         expected_series = Series(
-            expected,
+            expected_values,
             dtype="datetime64[ns]",
         )
         tm.assert_series_equal(result_series, expected_series)

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -957,26 +957,41 @@ class TestToDatetime:
         expected = Timestamp("20130101 00:00:00")
         assert result == expected
 
-    @pytest.mark.parametrize("datetime_likes, expected", [
+    @pytest.mark.parametrize(
+        "datetimelikes, expected",
+        [
+            (
+                [None] + [NaT] * start_caching_at + [np.nan],
+                [NaT] * (start_caching_at + 2),
+            ),
             (
                 [None] + [NaT] * start_caching_at + [Timestamp("2012-07-26")],
                 [NaT] * (start_caching_at + 1) + [Timestamp("2012-07-26")],
             ),
             (
-                [None] + [NaT] * start_caching_at + [Timestamp("2012-07-26"), "2012 July 26", "2012-07-26"],
-                [NaT] * (start_caching_at + 1) + [Timestamp("2012-07-26"), Timestamp("2012-07-26"), Timestamp("2012-07-26")],
-            )
-        ]
+                [None]
+                + [NaT] * start_caching_at
+                + [
+                    "2012 July 26",
+                    Timestamp("2012-07-26"),
+                ],
+                [NaT] * (start_caching_at + 1)
+                + [
+                    Timestamp("2012-07-26"),
+                    Timestamp("2012-07-26"),
+                ],
+            ),
+        ],
     )
-    def test_convert_object_to_datetime_with_cache(self, datetime_likes, expected):
+    def test_convert_object_to_datetime_with_cache(self, datetimelikes, expected):
         # GH#39882
         ser = Series(
-            datetime_likes,
+            datetimelikes,
             dtype="object",
         )
         result = to_datetime(ser, errors="coerce")
         expected = Series(
-            
+            expected,
             dtype="datetime64[ns]",
         )
         tm.assert_series_equal(result, expected)

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -957,15 +957,26 @@ class TestToDatetime:
         expected = Timestamp("20130101 00:00:00")
         assert result == expected
 
-    def test_convert_object_to_datetime_with_cache(self):
+    @pytest.mark.parametrize("datetime_likes, expected", [
+            (
+                [None] + [NaT] * start_caching_at + [Timestamp("2012-07-26")],
+                [NaT] * (start_caching_at + 1) + [Timestamp("2012-07-26")],
+            ),
+            (
+                [None] + [NaT] * start_caching_at + [Timestamp("2012-07-26"), "2012 July 26", "2012-07-26"],
+                [NaT] * (start_caching_at + 1) + [Timestamp("2012-07-26"), Timestamp("2012-07-26"), Timestamp("2012-07-26")],
+            )
+        ]
+    )
+    def test_convert_object_to_datetime_with_cache(self, datetime_likes, expected):
         # GH#39882
         ser = Series(
-            [None] + [NaT] * start_caching_at + [Timestamp("2012-07-26")],
+            datetime_likes,
             dtype="object",
         )
         result = to_datetime(ser, errors="coerce")
         expected = Series(
-            [NaT] * (start_caching_at + 1) + [Timestamp("2012-07-26")],
+            
             dtype="datetime64[ns]",
         )
         tm.assert_series_equal(result, expected)

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -8,6 +8,7 @@ from datetime import (
 )
 from decimal import Decimal
 import locale
+from typing import Any, List, Union
 
 from dateutil.parser import parse
 from dateutil.tz.tz import tzoffset
@@ -958,30 +959,23 @@ class TestToDatetime:
         assert result == expected
 
     @pytest.mark.parametrize(
-        "datetimelikes, expected",
-        [
+        "datetimelikes,expected",
+        (
             (
-                [None] + [NaT] * start_caching_at + [np.nan],
-                [NaT] * (start_caching_at + 2),
+                (None, np.nan) + (NaT,) * start_caching_at,
+                (NaT,) * (start_caching_at + 2),
             ),
             (
-                [None] + [NaT] * start_caching_at + [Timestamp("2012-07-26")],
-                [NaT] * (start_caching_at + 1) + [Timestamp("2012-07-26")],
+                (None, Timestamp("2012-07-26")) + (NaT,) * start_caching_at,
+                (NaT, Timestamp("2012-07-26")) + (NaT,) * start_caching_at,
             ),
             (
-                [None]
-                + [NaT] * start_caching_at
-                + [
-                    "2012 July 26",
-                    Timestamp("2012-07-26"),
-                ],
-                [NaT] * (start_caching_at + 1)
-                + [
-                    Timestamp("2012-07-26"),
-                    Timestamp("2012-07-26"),
-                ],
+                (None, "2012 July 26", Timestamp("2012-07-26"))
+                + (NaT,) * start_caching_at,
+                (NaT, Timestamp("2012-07-26"), Timestamp("2012-07-26"))
+                + (NaT,) * start_caching_at,
             ),
-        ],
+        ),
     )
     def test_convert_object_to_datetime_with_cache(self, datetimelikes, expected):
         # GH#39882
@@ -989,12 +983,12 @@ class TestToDatetime:
             datetimelikes,
             dtype="object",
         )
-        result = to_datetime(ser, errors="coerce")
-        expected = Series(
+        result_series = to_datetime(ser, errors="coerce")
+        expected_series = Series(
             expected,
             dtype="datetime64[ns]",
         )
-        tm.assert_series_equal(result, expected)
+        tm.assert_series_equal(result_series, expected_series)
 
     @pytest.mark.parametrize(
         "date, format",

